controllers/webhooks/workloads/subnamespaceanchor_validation_test.go

```diff
@@ -324,41 +351,183 @@ var _ = Describe("SubnamespaceanchorValidation", func() {
 				})
+
+				It("releases the lock on the old name", func() {
+					Expect(spaceNameRegistry.UnlockNameCallCount()).To(Equal(1))
+				})
 			})
 
-			When("the space name hasn't changed", func() {
+			When("registering the new name fails", func() {
 				BeforeEach(func() {
-					newAnchor.Labels[workloads.SpaceNameLabel] = "my-space"
-					newAnchor.Labels["something"] = "else"
-					listResult = []hnsv1alpha2.SubnamespaceAnchor{*anchor}
+					spaceNameRegistry.RegisterNameReturns(errors.New("boom!"))
 				})
 
-				It("succeeds", func() {
+				It("denies the request", func() {
+					Expect(response.Allowed).To(BeFalse())
+				})
+
+				It("releases the lock on the old name", func() {
+					Expect(spaceNameRegistry.UnlockNameCallCount()).To(Equal(1))
+				})
+			})
+		})
+	})
+
+	Describe("subnamespaceanchor deletion", func() {
+		JustBeforeEach(func() {
+			anchorJSON, err := json.Marshal(anchor)
+			Expect(err).NotTo(HaveOccurred())
+			request = admission.Request{
+				AdmissionRequest: admissionv1.AdmissionRequest{
+					Name:      anchor.Name,
+					Namespace: namespace,
+					Operation: admissionv1.Delete,
+					OldObject: runtime.RawExtension{
+						Raw: anchorJSON,
+					},
+				},
+			}
+			response = validatingWebhook.Handle(ctx, request)
+		})
+
+		Context("orgs", func() {
+			BeforeEach(func() {
+				anchor = &hnsv1alpha2.SubnamespaceAnchor{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      uuid.NewString(),
+						Namespace: namespace,
+						Labels: map[string]string{
+							workloads.OrgNameLabel: "my-org",
+						},
+					},
+				}
+			})
+
+			It("removes the name from the registry", func() {
+				Expect(orgNameRegistry.DeregisterNameCallCount()).To(Equal(1))
+				_, actualNamespace, name := orgNameRegistry.DeregisterNameArgsForCall(0)
+				Expect(actualNamespace).To(Equal(namespace))
+				Expect(name).To(Equal("my-org"))
+			})
+
+			When("deregistering fails", func() {
+				BeforeEach(func() {
+					orgNameRegistry.DeregisterNameReturns(errors.New("boom!"))
+				})
+
+				It("denies the deletion", func() {
+					Expect(response.Allowed).To(BeFalse())
+				})
+
+				When("the failure is a not found error", func() {
+					BeforeEach(func() {
+						orgNameRegistry.DeregisterNameReturns(k8serrors.NewNotFound(schema.GroupResource{}, "jim"))
+					})
+
+					It("allows the deletion", func() {
+						Expect(response.Allowed).To(BeTrue())
+					})
+				})
+			})
+
+			When("the anchor has no org or space label", func() {
+				BeforeEach(func() {
+					delete(anchor.Labels, workloads.OrgNameLabel)
+				})
+
+				It("allows the deletion anyway", func() {
+					Expect(response.Allowed).To(BeTrue())
+				})
+			})
+
+			When("the org has a space label as well", func() {
+				BeforeEach(func() {
+					anchor.Labels[workloads.SpaceNameLabel] = "my-space"
+				})
+
+				It("allows the deletion anyway", func() {
 					Expect(response.Allowed).To(BeTrue())
 				})
+
+				It("does not attempt to deregister any names", func() {
```

Comment:
We should probably have the `does not attempt to deregister any names` test in the previous context (`the anchor has no org or space labels`)

Reply:
Thanks! We've added some more checks for not calling the registry in the failure cases.